nodeup/pkg/model/kube_apiserver_healthcheck.go

```diff
@@ -0,0 +1,203 @@
+	userID := wellknownusers.KubeApiserverHealthcheckID
+	userName := wellknownusers.KubeApiserverHealthcheckName
+
+	// We create user a user and hardcode its UID to 10012 as
+	// that is the ID used inside the container.
+	{
+		c.AddTask(&nodetasks.UserTask{
+			Name:  userName,
+			UID:   userID,
+			Shell: "/sbin/nologin",
+			Home:  secretsDir,
+		})
+	}
+
+	clientKey, clientCert, err := b.buildClientKeypair(id)
+	if err != nil {
+		return err
+	}
+
+	c.AddTask(&nodetasks.File{
+		Path: filepath.Join(secretsDir),
+		Type: nodetasks.FileType_Directory,
+		Mode: s("0755"),
+	})
+
+	clientCertBytes, err := clientCert.AsBytes()
+	if err != nil {
+		return err
+	}
+	c.AddTask(&nodetasks.File{
+		Path:     filepath.Join(secretsDir, "client.crt"),
+		Contents: fi.NewBytesResource(clientCertBytes),
+		Type:     nodetasks.FileType_File,
+		Mode:     s("0644"),
+		Owner:    s(userName),
+	})
+
+	clientKeyBytes, err := clientKey.AsBytes()
+	if err != nil {
+		return err
+	}
+	c.AddTask(&nodetasks.File{
+		Path:     filepath.Join(secretsDir, "client.key"),
+		Contents: fi.NewBytesResource(clientKeyBytes),
+		Type:     nodetasks.FileType_File,
+		Mode:     s("0600"),
+		Owner:    s(userName),
+	})
+
+	cert, err := b.GetCert(fi.CertificateId_CA)
+	if err != nil {
+		return err
+	}
+
+	c.AddTask(&nodetasks.File{
+		Path:     filepath.Join(secretsDir, "ca.crt"),
+		Contents: fi.NewBytesResource(cert),
+		Type:     nodetasks.FileType_File,
+		Mode:     s("0644"),
+		Owner:    s(userName),
+	})
+
+	return nil
+}
+
+func (b *KubeAPIServerBuilder) buildClientKeypair(commonName string) (*pki.PrivateKey, *pki.Certificate, error) {
+	signerID := fi.CertificateId_CA
+
+	var signerKey *pki.PrivateKey
+	{
+		k, err := b.KeyStore.FindPrivateKey(signerID)
+		if err != nil {
+			return nil, nil, err
+		}
+		if k == nil {
+			return nil, nil, fmt.Errorf("private key %q not found", signerID)
+		}
+		signerKey = k
+	}
+
+	var signerCertificate *pki.Certificate
+	{
+		cert, err := b.KeyStore.FindCert(signerID)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if cert == nil {
+			return nil, nil, fmt.Errorf("certificate %q not found", signerID)
+		}
+
+		signerCertificate = cert
+	}
+
+	privateKey, err := pki.GeneratePrivateKey()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	template := &x509.Certificate{
+		BasicConstraintsValid: true,
+		IsCA:                  false,
+
+		Subject: pkix.Name{
+			CommonName: commonName,
+		},
+	}
+
+	// https://tools.ietf.org/html/rfc5280#section-4.2.1.3
+	//
+	// Digital signature allows the certificate to be used to verify
+	// digital signatures used during TLS negotiation.
+	template.KeyUsage = template.KeyUsage | x509.KeyUsageDigitalSignature
+	// KeyEncipherment allows the cert/key pair to be used to encrypt
+	// keys, including the symmetric keys negotiated during TLS setup
+	// and used for data transfer.
+	template.KeyUsage = template.KeyUsage | x509.KeyUsageKeyEncipherment
```

Comment:
By the way, `x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment` is the default for `pki.SignNewCertificate`.

Reply:
Ah - cool.  I suggest leaving it for now?  I think this function and the similar one to generate a kubeconfig end up being utils.

You can also see the beginnings of a thought I'm exploring here, which is putting less logic into nodeup.  The only parameter here is really the path and the CN, we could imagine this just being in the NodeUpConfig (or similar), meaning we would have more insight into exactly what nodeup does (because it would do less)

Reply:
Can leave for now. I'll look into extracting a function.

Reply:
Thanks - I don't think it's urgent, I just figured that's one way we could go, so it might not matter too much if we only have two of these (kubeconfig & "raw")